source/extensions/formatter/cel/cel.cc

```diff
@@ -52,15 +53,63 @@ CELFormatter::formatValueWithContext(const Envoy::Formatter::HttpFormatterContex
   return ValueUtil::stringValue(result.value());
 }
 
+TypedCELFormatter::TypedCELFormatter(const ::Envoy::LocalInfo::LocalInfo& local_info,
+                                     Expr::BuilderInstanceSharedPtr expr_builder,
+                                     const google::api::expr::v1alpha1::Expr& input_expr,
+                                     absl::optional<size_t>& max_length)
+    : local_info_(local_info), expr_builder_(expr_builder), parsed_expr_(input_expr),
+      max_length_(max_length) {
+  compiled_expr_ = Expr::createExpression(expr_builder_->builder(), parsed_expr_);
+}
+
+absl::optional<std::string>
+TypedCELFormatter::formatWithContext(const Envoy::Formatter::HttpFormatterContext& context,
+                                     const StreamInfo::StreamInfo& stream_info) const {
+  Protobuf::Arena arena;
+  auto eval_status =
+      Expr::evaluate(*compiled_expr_, arena, &local_info_, stream_info, &context.requestHeaders(),
+                     &context.responseHeaders(), &context.responseTrailers());
+  if (!eval_status.has_value() || eval_status.value().IsError()) {
+    return absl::nullopt;
+  }
+  const auto result = Expr::print(eval_status.value());
+  if (max_length_) {
+    return result.substr(0, max_length_.value());
+  }
+
+  return result;
+}
+
+ProtobufWkt::Value
+TypedCELFormatter::formatValueWithContext(const Envoy::Formatter::HttpFormatterContext& context,
+                                          const StreamInfo::StreamInfo& stream_info) const {
+  Protobuf::Arena arena;
+  auto eval_status =
+      Expr::evaluate(*compiled_expr_, arena, &local_info_, stream_info, &context.requestHeaders(),
+                     &context.responseHeaders(), &context.responseTrailers());
+  if (!eval_status.has_value() || eval_status.value().IsError()) {
+    return ValueUtil::nullValue();
+  }
+
+  ProtobufWkt::Value json_value;
+  if (!ExportAsProtoValue(eval_status.value(), &json_value).ok()) {
```

Comment:
@kyessenov you added this API to cel-cpp many moons ago, and it was used for a time in Envoy for WASM, but then it stopped being used at some point.

If you have any advice on this API being suitable or unsuitable, please let me know.